explcheck/src/explcheck.lua

```diff
@@ -0,0 +1,168 @@
+local common = require("explcheck-common")
+local preprocessing = require("explcheck-preprocessing")
+-- local lexical_analysis = require("explcheck-lexical-analysis")
+-- local syntactic_analysis = require("explcheck-syntactic-analysis")
+-- local semantic_analysis = require("explcheck-semantic-analysis")
+-- local pseudo_flow_analysis = require("explcheck-pseudo-flow-analysis")
+
+-- Transform a singular into plural if the count is zero or greater than two.
+local function pluralize(singular, count)
+  if count == 1 then
+    return singular
+  else
+    return singular .. "s"
+  end
+end
+
+-- Colorize a string using ASCII color codes.
+local function colorize(text, ...)
+  local buffer = {}
+  for _, color_code in ipairs({...}) do
+    table.insert(buffer, "\27[")
+    table.insert(buffer, tostring(color_code))
+    table.insert(buffer, "m")
+  end
+  table.insert(buffer, text)
+  table.insert(buffer, "\27[0m")
+  return table.concat(buffer, "")
+end
+
+-- Convert a byte number in a file to a line and column number in a file.
+local function convert_byte_to_line_and_column(state, byte_number)
+  local line_number = 0
+  for _, line_starting_byte_number in ipairs(state.line_starting_byte_numbers) do
+    if line_starting_byte_number > byte_number then
+      break
+    end
+    line_number = line_number + 1
+  end
+  assert(line_number > 0)
+  local line_starting_byte_number = state.line_starting_byte_numbers[line_number]
+  assert(line_starting_byte_number <= byte_number)
+  local column_number = byte_number - line_starting_byte_number + 1
+  return line_number, column_number
+end
+
+-- Print warnings and errors after analyzing a file.
+local function print_warnings_and_errors(state, is_last_file)
```

Comment:
Pass `errors`, `warnings`, and `ignored_issues` directly to reduce the spread of the `state` god object.

Reply:
Fixed in commit 82e0dee2cd02c1d53633cddd42ab8fbd03f1d9f3 from this PR.